src/ir/auth_logic/souffle_emitter.h

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2022 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SRC_IR_AUTH_LOGIC_SOUFFLE_EMITTER_H_
+#define SRC_IR_AUTH_LOGIC_SOUFFLE_EMITTER_H_
+
+#include "absl/container/btree_set.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_join.h"
+#include "src/ir/auth_logic/datalog_ir.h"
+
+namespace raksha::ir::auth_logic {
+
+class SouffleEmitter {
+ public:
+  static std::string EmitProgram(const DLIRProgram& program) {
+    SouffleEmitter emitter;
+    std::string body = emitter.EmitProgramBody(program);
+    std::string outputs = emitter.EmitOutputs(program);
+    std::string decls = emitter.EmitDeclarations();
+    return absl::StrCat(std::move(body), "\n", std::move(outputs), "\n",
+                        std::move(decls), "\n");
+  }
+
+ private:
+  SouffleEmitter() : declarations_(absl::btree_set<Predicate>()){};
```

Comment:
I'm pretty sure `declarations_`'s default ctor creates an empty set as you would expect. Therefore, you can just write this as `declarations_` or omit it altogether and just have `SouffleEmitter() { }` or (I think) even `SouffleEmitter() = default;`.